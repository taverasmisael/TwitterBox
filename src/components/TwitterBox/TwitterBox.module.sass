$twitterColor: #039ff5;
$twitterSoft: rgba($twitterColor, 0.4);
$profileAccent: var(--profile-accent, $twitterSoft);
.container
  align-items: center
  background-color: $profileAccent
  color: var(--profile-accent, #fff)
  display: grid
  grid-template-areas: "pic textarea" ". actions"
  grid-template-columns: 36px 1fr
  grid-template-rows: 1fr
  grid-gap: 6px
  padding: 10px 12px
  .textarea
    max-height: 36px
  &.containerIsOpen
    align-items: flex-start
    min-height: 150px
    .textarea
      min-height: 100%
      max-height: 100%
      &:focus
        border: 2px solid rgba($twitterColor, 0.6)
    .actions
      display: grid
      grid-template-areas: "addons . submit"
      grid-template-rows: 1fr
      grid-template-columns: 4fr 1fr 1fr

.profilePic
  border-radius: 50%
  display: block
  grid-area: pic
  height: auto
  width: 100%

.textarea
  border: 0
  border-radius: 4px
  box-shadow: none
  grid-area: textarea
  font-size:  18px
  height: 100%
  padding: 8px
  resize: none
  will-change: border, max-height
  width: 100%
  &:focus
    border-radius: 4px
    outline: none

.actions
  display: none
  grid-area: actions

.submit
  grid-area: submit