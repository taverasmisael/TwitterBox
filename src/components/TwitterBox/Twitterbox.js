// @flow
import React, { PureComponent } from 'react'
import Textarea from 'react-textarea-autosize'
import classnames from 'classnames'
import styles from './TwitterBox.module.sass'

type Props = {
  profile: {
    name: string,
    handler: string,
    url: string,
    picture: string,
  },
}

type State = {
  isOpen: boolean,
  canTweet: boolean,
  tweetContent: string,
  message: string,
}

class TwitterBox extends PureComponent<Props, State> {
  static defaultProps = {
    profile: {},
  }

  state = {
    isOpen: false,
    canTweet: false,
    tweetContent: '',
    message: '',
  }
  maxTweetLength = 10

  render() {
    const { profile } = this.props
    const { isOpen, canTweet, tweetContent, message } = this.state
    return (
      <div
        className={classnames(styles.container, {
          [styles.containerIsOpen]: isOpen,
        })}
      >
        <img
          alt="Profile"
          src={profile.picture}
          className={styles.profilePic}
        />
        <Textarea
          type="text"
          aria-label="Write your tweet message"
          data-gramm="false"
          placeholder="What is on your mind?"
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onChange={this.onChange}
          value={tweetContent}
          className={styles.textarea}
        />
        <div className={styles.actions}>
          <span
            className={classnames(styles.message, {
              [styles.messageError]: !canTweet,
              [styles.messageVisible]: tweetContent,
            })}
          >
            {message}
          </span>
          <button
            disabled={!canTweet}
            type="button"
            className={`button is-primary is-rounded ${styles.submit}`}
          >
            Tweet
          </button>
        </div>
      </div>
    )
  }

  onFocus = () => this.setState({ isOpen: true })
  onBlur = () => this.setState(state => ({ isOpen: state.tweetContent }))
  onChange = ({ target }) => {
    const canTweet =
      target.value.length && target.value.length <= this.maxTweetLength
    const message = canTweet
      ? `You have ${target.value.length} out of ${this.maxTweetLength} left`
      : `You have ${target.value.length -
          this.maxTweetLength} characters above the limit(${
          this.maxTweetLength
        })`
    return this.setState({
      tweetContent: target.value,
      canTweet,
      message,
    })
  }
}

export default TwitterBox
