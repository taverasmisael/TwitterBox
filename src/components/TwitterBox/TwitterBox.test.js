// @flow
import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import TwitterBox from './TwitterBox'

describe('component/TwitterBox', () => {
  const profile = {
    name: 'Misael Taveras',
    handler: '@taverasmisael',
    url: 'https://twitter.com/taverasmisael',
    picture:
      'https://pbs.twimg.com/profile_images/926912434610114560/MAMgQ07A_bigger.jpg',
  }

  describe('core and non-open state', () => {
    it('should match snapshot', () => {
      const { container } = render(<TwitterBox />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should only display profile image and input when closed', () => {
      const { getByAltText, getByLabelText } = render(<TwitterBox />)
      expect(getByAltText('Profile')).toBeInTheDocument()
      expect(getByLabelText('Write your tweet message')).toBeInTheDocument()
    })

    it('should render the right profile image', () => {
      const { getByAltText } = render(<TwitterBox profile={profile} />)
      expect(getByAltText('Profile').src).toBe(profile.picture)
    })

    it('should expand to a full version when the input has focus', () => {
      const { getByLabelText } = render(<TwitterBox />)
      const twitterBox = getByLabelText('Write your tweet message')
      fireEvent.focus(twitterBox)
    })
  })

  describe('functionality', () => {
    describe('check tweet length', () => {
      it('disable submit if tweet length is 0', () => {
        const { getByText } = render(<TwitterBox />)
        const submitBtn = getByText('Tweet')
        expect(submitBtn.disabled).toBe(true)
      })
      it('enables submit if tweet is less than maxLen', () => {
        const { getByLabelText, getByText } = render(<TwitterBox />)
        const tweetInput = getByLabelText('Write your tweet message')
        const submitBtn = getByText('Tweet')
        fireEvent.change(tweetInput, { target: { value: 'Hi Tweet' } })
        expect(submitBtn.disabled).toBe(false)
      })

      it('disables submit if tweet is more than maxLen', () => {
        const { getByLabelText, getByText } = render(<TwitterBox />)
        const tweetInput = getByLabelText('Write your tweet message')
        const submitBtn = getByText('Tweet')
        fireEvent.change(tweetInput, { target: { value: 'Hi Tweet' } })
        expect(submitBtn.disabled).toBe(false)
        fireEvent.change(tweetInput, {
          target: { value: 'Hi super mega long Tweet' },
        })
        expect(submitBtn.disabled).toBe(true)
      })

      it('disables submit if tweet goes back to 0 characters', () => {
        const { getByLabelText, getByText } = render(<TwitterBox />)
        const tweetInput = getByLabelText('Write your tweet message')
        const submitBtn = getByText('Tweet')
        fireEvent.change(tweetInput, { target: { value: 'Hi Tweet' } })
        expect(submitBtn.disabled).toBe(false)
        fireEvent.change(tweetInput, { target: { value: '' } })
        expect(submitBtn.disabled).toBe(true)
      })
    })

    describe('tweet length feedback', () => {
      it('tells me how many characters I got left', () => {
        const { getByLabelText, getByText } = render(<TwitterBox />)
        const tweetInput = getByLabelText('Write your tweet message')
        fireEvent.change(tweetInput, { target: { value: 'Hi Tweet' } })
        expect(getByText(/You have \d* out of \d* left/).textContent).toEqual(
          'You have 8 out of 10 left'
        )
      })

      it('tells me when I have exceded the maxLength', () => {
        const { getByLabelText, getByText } = render(<TwitterBox />)
        const tweetInput = getByLabelText('Write your tweet message')
        fireEvent.change(tweetInput, { target: { value: 'Hi super long Tweet' } })
        expect(getByText(/You have \d* characters above the limit(\d*)/).textContent).toEqual(
          'You have 9 characters above the limit(10)'
        )
      })
    })
  })
})
