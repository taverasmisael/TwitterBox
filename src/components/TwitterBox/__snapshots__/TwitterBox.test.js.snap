// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/TwitterBox core and non-open state should match snapshot 1`] = `
<div
  class="container"
>
  <img
    alt="Profile"
    class="profilePic"
  />
  <textarea
    aria-label="Write your tweet message"
    class="textarea"
    data-gramm="false"
    placeholder="What is on your mind?"
    style="height: 0px;"
    type="text"
  />
  <div
    class="actions"
  >
    <span
      class="message"
    />
    <button
      class="button is-primary is-rounded submit"
      disabled=""
      type="button"
    >
      Tweet
    </button>
  </div>
</div>
`;
